src/core/ipe-setup.c

```diff
@@ -0,0 +1,122 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "conf-files.h"
+#include "copy.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "ipe-setup.h"
+#include "path-util.h"
+
+#define IPE_SECFS_DIR "/sys/kernel/security/ipe"
+#define IPE_SECFS_NEW_POLICY IPE_SECFS_DIR "/new_policy"
+#define IPE_SECFS_POLICIES IPE_SECFS_DIR "/policies/"
+
+int ipe_setup(void) {
+#if ENABLE_IPE
+        _cleanup_strv_free_ char **policies = NULL;
+        int r;
+
+        /* Very quick smoke tests first: this is in the citical, sequential boot path, and in most cases it
+         * is unlikely this will be configured, so do the fastest existence checks first and immediately
+         * return if there's nothing to do. */
+
+        if (access(IPE_SECFS_DIR, F_OK) < 0) {
+                log_debug_errno(errno, "IPE support is disabled in the kernel, ignoring: %m");
+                return 0;
+        }
+
+        bool found = false;
+        STRV_FOREACH(p, CONF_PATHS_NULSTR("ipe"))
+                if (dir_is_empty(p, /* ignore_hidden_or_backup= */ true) == 0) {
+                        found = true;
+                        break;
+                }
+        if (!found) {
+                log_debug("No IPE policy files found, ignoring.");
+                return 0;
+        }
+
+        r = conf_files_list_nulstr(
+                        &policies,
```

Comment:
Hmm, why _nulstr? It does strv_split_nulstr under the hood, i.e. meaningless allocation?

Reply:
drepper's ELF/shared library paper suggests to use "nulstrs" over pointer arrays, for efficiency. not sure how much this really matters, but you save a bunch of relocations that way.

https://www.cs.dartmouth.edu/~sergey/cs108/ABI/UlrichDrepper-How-To-Write-Shared-Libraries.pdf

Reply:
Well, the thing is that conf_files_list_nulstr() splits it into strv anyways...

Reply:
> Well, the thing is that conf_files_list_nulstr() splits it into strv anyways...

that is true.

The only difference in effect is that on non-IPE systems you'll save a bunch of relocations by using nulstr instead of strv here. You'd buy this by doing some extra malloc/free cycles in case IPE is actually enabled.

Reply:
This part needs to be optimized for non-IPE systems, which is going to be 99.999% of the cases